File: cloud-admin-portal/backend/app/session.py
"""Tenant session management.

Tenant sessions store validated credentials for user management operations.
This avoids passing passwords in URLs and reduces credential exposure.
"""

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Dict, Optional
import secrets

from .config import get_settings


@dataclass
class TenantSession:
    """Stores validated tenant credentials for user management operations."""
    
    session_id: str
    tenant_id: str
    keyspace: str
    admin_user: str
    admin_password: str
    created_at: datetime
    expires_at: datetime
    
    def is_expired(self) -> bool:
        return datetime.now(timezone.utc) > self.expires_at


class SessionManager:
    """Manages tenant sessions with automatic cleanup."""
    
    def __init__(self):
        self._sessions: Dict[str, TenantSession] = {}
    
    def create_session(
        self,
        tenant_id: str,
        keyspace: str,
        admin_user: str,
        admin_password: str
    ) -> TenantSession:
        settings = get_settings()
        
        session_id = f"ts_{secrets.token_hex(16)}"
        
        now = datetime.now(timezone.utc)
        expires_at = now + timedelta(hours=settings.session_ttl_hours)
        
        session = TenantSession(
            session_id=session_id,
            tenant_id=tenant_id,
            keyspace=keyspace,
            admin_user=admin_user,
            admin_password=admin_password,
            created_at=now,
            expires_at=expires_at,
        )
        
        self._sessions[session_id] = session
        self._cleanup_expired()
        
        return session
    
    def get_session(self, session_id: str) -> Optional[TenantSession]:
        session = self._sessions.get(session_id)
        
        if session is None:
            return None
        
        if session.is_expired():
            del self._sessions[session_id]
            return None
        
        return session
    
    def validate_session(self, session_id: str, tenant_id: str) -> Optional[TenantSession]:
        session = self.get_session(session_id)
        
        if session is None:
            return None
        
        if session.tenant_id != tenant_id:
            return None
        
        return session
    
    def delete_session(self, session_id: str) -> bool:
        """Delete a session.
        
        Args:
            session_id: Session ID
        
        Returns:
            True if session was deleted, False if not found
        """
        if session_id in self._sessions:
            del self._sessions[session_id]
            return True
        return False
    
    def clear_all(self):
        """Clear all sessions. Called on shutdown."""
        self._sessions.clear()
    
    def _cleanup_expired(self):
        """Remove expired sessions."""
        now = datetime.now(timezone.utc)
        expired = [
            sid for sid, session in self._sessions.items()
            if session.expires_at < now
        ]
        for sid in expired:
            del self._sessions[sid]


# Global session manager instance
session_manager = SessionManager()

File: cloud-admin-portal/backend/src/session.rs
use std::collections::HashMap;
use std::sync::RwLock;

use chrono::{Duration, Utc};
use rand::Rng;

#[derive(Clone, Debug)]
pub struct TenantSession {
    pub session_id: String,
    pub tenant_id: String,
    pub admin_user: String,
    pub admin_password: String,
    pub expires_at: String,
}

pub struct SessionManager {
    sessions: RwLock<HashMap<String, TenantSession>>,
    ttl_hours: i64,
}

impl SessionManager {
    pub fn new(ttl_hours: u64) -> Self {
        Self {
            sessions: RwLock::new(HashMap::new()),
            ttl_hours: ttl_hours as i64,
        }
    }

    pub fn create_session(
        &self,
        tenant_id: &str,
        admin_user: &str,
        admin_password: &str,
    ) -> TenantSession {
        let session_id = generate_session_id();
        let expires_at = (Utc::now() + Duration::hours(self.ttl_hours)).to_rfc3339();

        let session = TenantSession {
            session_id: session_id.clone(),
            tenant_id: tenant_id.to_string(),
            admin_user: admin_user.to_string(),
            admin_password: admin_password.to_string(),
            expires_at,
        };

        if let Ok(mut map) = self.sessions.write() {
            map.insert(session_id, session.clone());
        }
        session
    }

    pub fn validate_session(&self, session_id: &str, tenant_id: &str) -> Option<TenantSession> {
        let map = self.sessions.read().ok()?;
        let session = map.get(session_id)?;

        if session.tenant_id != tenant_id {
            return None;
        }

        if let Ok(expires) = chrono::DateTime::parse_from_rfc3339(&session.expires_at) {
            if Utc::now() > expires {
                return None;
            }
        }

        Some(session.clone())
    }

    pub fn clear_all(&self) {
        if let Ok(mut map) = self.sessions.write() {
            map.clear();
        }
    }
}

fn generate_session_id() -> String {
    let mut bytes = [0u8; 32];
    rand::thread_rng().fill(&mut bytes);
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}
